### PR TITLE
lumberjack changes to runWithRetry

### PR DESCRIPTION
**Pull request by appym31**
Adding lumberjack to the runWithRetry method

### CODE REVIEW

**Review by tanviraumi — COMMENTED** [review]
@hedasilv and @znewton Is Lumberjack supposed to run client and service side? I think server-service-core is being used in "@fluidframework/local-driver"

**Comment by hedasilv** [comment]
> @hedasilv and @znewton Is Lumberjack supposed to run client and service side? I think server-service-core is being used in "@fluidframework/local-driver"

Good question @tanviraumi. I think Lumberjack itself (excluding the Lumberjack engine) is isomorphic. Lumberjack and Lumber types are pure typescript, and only depend on a few external packages: 
```
... (from package.json in services-telemetry)
"dependencies": {
    "@fluidframework/common-utils": "^0.32.1",
    "json-stringify-safe": "^5.0.1",
    "path-browserify": "^1.0.1",
    "uuid": "^8.3.1"
  },
```

In fact, #7057 made Lumberjack "browser friendly" by using path-browserify. Also, in #6868, server-services-telemetry was added as a dependency to server-local-server, which I believe was a requirement to make some unit tests work.

I believe we should have no trouble. How can we test this? @appym31 , would you be able to test "@fluidframework/local-driver" to make sure things run smoothly?

File: server/routerlicious/packages/services-core/src/runWithRetry.ts [hunk 1]
```diff
@@ -12,6 +13,7 @@ export async function runWithRetry<T>(
     maxRetries: number,
     retryAfterMs: number,
     logger?: ILogger,
+    lumberjack?: typeof Lumberjack,
```
Comment (hedasilv) [inline server/routerlicious/packages/services-core/src/runWithRetry.ts:16]: Why do we have to make Lumberjack as a parameter?
Can we just reference the global Lumberjack instead?

**Comment by znewton** [comment]
It looks like @pradeepvairamani added Lumberjack setup to local-server (and driver) in #6868, which is what allowed #6944 and #7395 to also use global Lumberjack within the isomorphic lambdas. It appears that the local-server currently hardcodes the only Lumberjack engine to be `TestEngine1` which pipes the logs to unto the void.

**Comment by appym31** [comment]
> > @hedasilv and @znewton Is Lumberjack supposed to run client and service side? I think server-service-core is being used in "@fluidframework/local-driver"
> 
> Good question @tanviraumi. I think Lumberjack itself (excluding the Lumberjack engine) is isomorphic. Lumberjack and Lumber types are pure typescript, and only depend on a few external packages:
> 
> ```
> ... (from package.json in services-telemetry)
> "dependencies": {
>     "@fluidframework/common-utils": "^0.32.1",
>     "json-stringify-safe": "^5.0.1",
>     "path-browserify": "^1.0.1",
>     "uuid": "^8.3.1"
>   },
> ```
> 
> In fact, #7057 made Lumberjack "browser friendly" by using path-browserify. Also, in #6868, server-services-telemetry was added as a dependency to server-local-server, which I believe was a requirement to make some unit tests work.
> 
> I believe we should have no trouble. Especially since we are only adding Lumberjack to `runWithRetry()` within server-services-core. @appym31, do we ever call `runWithRetry` in "@fluidframework/local-driver"?
> 
> If yes, how can we test this in "@fluidframework/local-driver" to make sure things run smoothly?

On @hedasilv 's question, we do not call runWithRetry from services-core in local-driver.

File: server/routerlicious/packages/services-core/src/runWithRetry.ts [hunk 2]
```diff
@@ -26,14 +27,17 @@ export async function runWithRetry<T>(
...
                 break;
             } else if (shouldRetry !== undefined && shouldRetry(error) === false)
             {
                 logger?.info(`Should not retry ${callName}`);
+                Lumberjack.info(`Should not retry ${callName}`);
                 return Promise.reject(error);
             }
             if (retryCount >= maxRetries) {
                 logger?.info(`Error after retrying ${retryCount} times, rejecting`);
+                Lumberjack.info(`Error after retrying ${retryCount} times, rejecting`);
```
Comment (hedasilv) [inline server/routerlicious/packages/services-core/src/runWithRetry.ts:40]: Should this be `Lumberjack.error()` instead? Could also be the chance to change `logger?.info` to `logger?.error()`

File: server/routerlicious/packages/services-core/src/runWithRetry.ts [hunk 3]
```diff
@@ -26,14 +27,17 @@ export async function runWithRetry<T>(
             logger?.info(`Error running ${callName}: retryCount ${retryCount}, error ${error}`);
             if (shouldIgnoreError !== undefined && shouldIgnoreError(error) === true) {
                 logger?.info(`Should ignore error for ${callName}`);
+                Lumberjack.info(`Should ignore error for ${callName}`);
                 break;
             } else if (shouldRetry !== undefined && shouldRetry(error) === false)
             {
                 logger?.info(`Should not retry ${callName}`);
+                Lumberjack.info(`Should not retry ${callName}`);
```
Comment (hedasilv) [inline server/routerlicious/packages/services-core/src/runWithRetry.ts:35]: Should we also use `Lumberjack.error()` here? also, should we explicitly say "rejecting" at the end of the error message to make it clear the retry will not be executed and the call will fail?

File: server/routerlicious/packages/services-core/src/runWithRetry.ts [hunk 4]
```diff
@@ -26,14 +27,17 @@ export async function runWithRetry<T>(
             logger?.info(`Error running ${callName}: retryCount ${retryCount}, error ${error}`);
```
Comment (hedasilv) [inline server/routerlicious/packages/services-core/src/runWithRetry.ts:27]: Did we miss a Lumberjack here, or do we not want to add a Lumberjack log in this case?

File: server/routerlicious/packages/services-core/src/runWithRetry.ts (hunk 3 above)
Comment (hedasilv) [reply server/routerlicious/packages/services-core/src/runWithRetry.ts:35]: Another thing that could be helpful is saying `should not retry ${callName} for the current error.`. And then, pass the error object to `Lumberjack.error()`

**Review by hedasilv — APPROVED** [review]

File: server/routerlicious/packages/services-core/src/runWithRetry.ts [hunk 5]
```diff
@@ -22,18 +23,24 @@ export async function runWithRetry<T>(
         try {
             result = await api();
             success = true;
+            logger?.info(`Succeeded in executing ${callName} with ${retryCount} retries`);
+            Lumberjack.info(`Succeeded in executing ${callName} with ${retryCount} retries`);
```
Comment (znewton) [inline server/routerlicious/packages/services-core/src/runWithRetry.ts:27]: Maybe we should only log this if `retryCount >= 1`?

Comment (pradeepvairamani) [reply server/routerlicious/packages/services-core/src/runWithRetry.ts:27]: Yes please. This log will be very noisy otherwise.

**Review by znewton — APPROVED** [review]

**Review by hedasilv — APPROVED** [review]

**Pull request merged**